### PR TITLE
MM-20104 -Migrate channelCache cache from store/sqlstore/channel_store.go to th…

### LINKED ISSUE #13101: Migrate channelCache cache from store/sqlstore/channel_store.go to the new store/localcachelayer
Mattermost is migrating the old way to implement caches moving them behind a transparent layer on to of the store and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `channelCache` cache variable in the `store/sqlstore/channel_store.go` file to the `store/localcachelayer`.

The expected way to implement it is, go to the `store/sqlstore/channel_store.go` file get all the code that is related to the cache variable `channelCache` and move it to the `store/localcachelayer`

Besides is expected to generate tests mocking the database store, and checking that the cache behavior is the expected one.

Example: mattermost/mattermost-server#13045

If you have any question, you can contact with `jesus.espino` at https://community.mattermost.com

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-20104

### PR DESCRIPTION
**Pull request by nadalfederer**
Summary
Migrate channelCache cache from store/sqlstore/channel_store.go to the new store/localcachelayer 

Ticket Link
https://mattermost.atlassian.net/browse/MM-20104

closes #13101

### CODE REVIEW

**Comment by nadalfederer** [comment]
@lindy65 hi, it's ready, I'm not sure whom to ask for a review

File: store/localcachelayer/channel_layer.go [hunk 1]
```diff
@@ -132,3 +149,32 @@ func (s LocalCacheChannelStore) GetPinnedPostCount(channelId string, allowFromCa
...
+	return s.get(id, false, allowFromCache)
+}
+
+func (s LocalCacheChannelStore) get(id string, master bool, allowFromCache bool) (*model.Channel, *model.AppError) {
+
+	if allowFromCache {
+		if cacheItem := s.rootStore.doStandardReadCache(s.rootStore.channelByIdCache, id); cacheItem != nil {
+			if s.rootStore.metrics != nil {
+				s.rootStore.metrics.IncrementMemCacheHitCounter("Channel")
+			}
```
Comment (streamer45) [inline store/localcachelayer/channel_layer.go]: The hit counter is already incremented by calling `rootStore.doStandardReadCache()`. We can safely remove this block.

**Comment by nadalfederer** [comment]
@streamer45 Thanks for the review. I fixed current bugs.

**Comment by lindy65** [comment]
Thanks for the PR@nadalfederer :)

@streamer45 - please let me know what exactly needs to be tested for this PR? I'm thinking its everything around channels?  If so, does this include archived channels?

**Comment by streamer45** [comment]
> Thanks for the PR@nadalfederer :)
> 
> @streamer45 - please let me know what exactly needs to be tested for this PR? I'm thinking its everything around channels? If so, does this include archived channels?

Anything that touches channel cache should be tested. Specifically:

- Channel deletion
- Channel update
- Channel privacy update 
- Post creation

**Comment by lindy65** [comment]
Hi @nadalfederer @streamer45 

I tested on the spinwick as per `streamer45's` test instructions above...

1. I'm seeing a "More..." option below PRIVATE CHANNELS whether I'm logged in as Sysadmin or User-1.  I don't see this option on `community-daily`.  Is this caused by something in this PR?

image

2. On leaving a private channel, I get a white screen with an error in the js console. This is fixed after a refresh and channel is left and removed from the left-hand-side channels list.

image

Other than the above, I did not find any other issues.

**Comment by Willyfrog** [comment]
@lindy65 there was a report for 2 i believe: https://mattermost.atlassian.net/browse/MM-20920 it's already fixed on master, updating and resolving conflicts should probably fix it.

**Comment by lindy65** [comment]
Thanks @Willyfrog 👍 I'll wait for @nadalfederer to update and resolve conflicts as well as look at 1) before testing again...

**Comment by nadalfederer** [comment]
@lindy65 I dont see your problem locally 1). All I dont have this button on either admin or common user
Screenshot from 2019-12-05 17-33-13
Screenshot from 2019-12-05 17-33-50

**Comment by nadalfederer** [comment]
Im sorry, I accidentally clicked "close and comment" button"

**Comment by lindy65** [comment]
Thanks @nadalfederer - I am still seeing the "More..." below PRIVATE CHANNELS on the spinwick. I've checked our `master.test.mattermost.com`  test server as well as `community-daily` and I don't see it there so I am not sure why I am seeing it on the spinwick.

@streamer45 @Willyfrog - would you have any idea why I am seeing it on this spinwick?

image

**Comment by streamer45** [comment]
> Thanks @nadalfederer - I am still seeing the "More..." below PRIVATE CHANNELS on the spinwick. I've checked our `master.test.mattermost.com` test server as well as `community-daily` and I don't see it there so I am not sure why I am seeing it on the spinwick.
> 
> @streamer45 @Willyfrog - would you have any idea why I am seeing it on this spinwick?
> 
> image

Looks like that has to do with the experimental setting to view archived channels being enabled. Not related to this PR.

**Comment by lindy65** [comment]
Thanks @streamer45!  You are absolutely correct :)  I added a licence to the server, switched off the "Allow users to view archived channels" setting and the "+" has disappeared from below PRIVATE CHANNELS 👍

File: store/localcachelayer/channel_layer.go [hunk 2]
```diff
@@ -132,3 +149,25 @@ func (s LocalCacheChannelStore) GetPinnedPostCount(channelId string, allowFromCa
 
 	return count, nil
 }
+
+func (s LocalCacheChannelStore) Get(id string, allowFromCache bool) (*model.Channel, *model.AppError) {
+	return s.get(id, false, allowFromCache)
```
Comment (Willyfrog) [inline store/localcachelayer/channel_layer.go]: why not adding `get` code directly here, i didn't see any other use, so this is adding layers without any benefit

File: store/localcachelayer/channel_layer.go [hunk 3]
```diff
@@ -132,3 +149,25 @@ func (s LocalCacheChannelStore) GetPinnedPostCount(channelId string, allowFromCa
...
+		}
+	}
+
+	ch, err := s.ChannelStore.Get(id, allowFromCache)
+
+	if allowFromCache && err == nil {
+		s.rootStore.doStandardAddToCache(s.rootStore.channelByIdCache, id, ch)
+	}
+
+	return ch, nil
```
Comment (Willyfrog) [inline store/localcachelayer/channel_layer.go]: this is ignoring `err`, shouldn't we at least log the error or pass it through?

**Comment by nadalfederer** [comment]
@Willyfrog thanks for the review. I fixed current issues

**Pull request merged**